src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotOverloadOperatorEqualsOnReferenceTypes.cs

```diff
@@ -0,0 +1,60 @@
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class DoNotOverloadOperatorEqualsOnReferenceTypes : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA1046";
+
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(MicrosoftCodeQualityAnalyzersResources.DoNotOverloadOperatorEqualsOnReferenceTypesTitle), MicrosoftCodeQualityAnalyzersResources.ResourceManager, typeof(MicrosoftCodeQualityAnalyzersResources));
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(MicrosoftCodeQualityAnalyzersResources.DoNotOverloadOperatorEqualsOnReferenceTypesMessage), MicrosoftCodeQualityAnalyzersResources.ResourceManager, typeof(MicrosoftCodeQualityAnalyzersResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftCodeQualityAnalyzersResources.DoNotOverloadOperatorEqualsOnReferenceTypesDescription), MicrosoftCodeQualityAnalyzersResources.ResourceManager, typeof(MicrosoftCodeQualityAnalyzersResources));
+
+        public static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Design,
+            RuleLevel.IdeSuggestion,
+            description: s_localizableDescription,
+            isPortedFxCopRule: true,
+            isDataflowRule: false,
+            isEnabledByDefaultInFxCopAnalyzers: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterSymbolAction(context =>
+            {
+                var method = (IMethodSymbol)context.Symbol;
+
+                // Method is not operator equality or not on a class
+                if (method.MethodKind != MethodKind.UserDefinedOperator ||
+                    method.Name != "op_Equality" ||
+                    method.ContainingType.TypeKind != TypeKind.Class)
+                {
+                    return;
+                }
+
+                // FxCop compat: only analyze externally visible symbols by default.
+                if (!method.MatchesConfiguredVisibility(context.Options, Rule, context.CancellationToken))
+                {
+                    return;
+                }
+
+                context.ReportDiagnostic(method.CreateDiagnostic(Rule, method.ContainingType.Name));
```

Comment:
There's a CONSIDER rule for overriding op_Equality for anything that implements `IEquatable<TheirOwnType>`.

CONSIDER not reporting a diagnostic in such cases :smile:.

Reply:
+1

Reply:
@bartonjs I might be missing something but the only guideline I found suggesting to override operator == on `IEquatable` was for value types. In here this is analyzing only reference type. Do you have any doc?

Reply:
_Framework Design Guidelines, 2nd ed._, 8.6 (IComparable\<T\> and IEquatable\<T\>), pp. 264

It does seem to be missing from the guidelines published to MSDN/docs.microsoft.com; when 3rd edition releases (later this year) we'll be updating the "MSDN" guidelines-only version to match.

Reply:
Thank you!